feat(curriculum): add string method workshop (#66936)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/workshop-employee-profile-validator/69e6284f3b7659c5a05c2a79.md

---
id: 69e6284f3b7659c5a05c2a79
title: Step 5
challengeType: 20
dashedName: step-5
---

# --description--

You can check whether a substring exists inside a string using the `in` keyword. It returns `True` if the substring exists and `False` otherwise. Here's an example:

```py
sentence = 'I love Python'
print('Python' in sentence) # True
print('Java' in sentence) # False
```

> Create a variable named `has_at` that checks if `'@'` is in `email`.

Reviewer: ```suggestion
Create a variable named `has_at` that checks if the character `@` is in `email`.
```

# --hints--

> You should create a variable named `has_at` and assign it the value `'@' in email`.

Reviewer 2: This hint's assert only runs `has_variable('has_at')`; the value is graded by the next hint. Trimming it also matches the wording used in steps 6, 10, 12, 14, 15, 17, and 20.

```suggestion
You should create a variable named `has_at`.
```

```js
({ test: () => assert(runPython(`_Node(_code).has_variable('has_at')`)) })
```

Your `has_at` variable should have the value `'@' in email`.

```js
({ test: () => assert(runPython(`
node = _Node(_code).find_variable('has_at')
not node.is_empty() and (
  node.is_equivalent("has_at = '@' in email")
)
`)) })
```

> 

Reviewer: ````suggestion

You should print the result of concatenating the string `Has @ symbol: ` and the `has_at` variable.

```js
({ test: () => assert(runPython(`_Node(_code).has_call("print('Has @ symbol: ' + str(has_at))")`)) })
```

````

---
id: 69e6284f3b7659c5a05c2a79
title: Step 5
challengeType: 20
dashedName: step-5
---

# --description--

You can check whether a substring exists inside a string using the `in` keyword. It returns `True` or `False`. For example:

```py
sentence = 'I love Python'
result = 'Python' in sentence
print(result)  # Output: True
```

Create two variables:

> 

Reviewer: Since this is the first time `in` is used, create one variable at a time. `has_com` and the related print should go in another step.

---
id: 69e6284f3b7659c5a05c2a79
title: Step 5
challengeType: 20
dashedName: step-5
---

# --description--

You can check whether a substring exists inside a string using the `in` keyword. It returns `True` if the substring exists and `False` otherwise. Here's an example:

```py
sentence = 'I love Python'
print('Python' in sentence) # True
print('Java' in sentence) # False
```

Create a variable named `has_at` that checks if the character `@` is in `email`.

> Then print `has_at` so that the output reads: `Has @ symbol: True`

Reviewer 2: L18 says the variable "checks" something. The variable holds the result of the check rather than performing it, and step 6 copies this phrasing, so both are worth rewording.

The trailing punctuation is the cross-step item: "so that the output reads" is followed by a colon and no period in steps 2, 3 and 5, a colon and a period in steps 6, 11, 13 and 16, and no colon with a period in steps 4, 9 and 15. Every hint line in the workshop uses the last of those, so aligning the descriptions to it is the smallest change.

```suggestion
Create a variable named `has_at` and assign it the result of checking whether the `@` character is in `email`.

Then print `has_at` so that the output reads `Has @ symbol: True`.
```

# --hints--

You should create a variable named `has_at`.

```js
({ test: () => assert(runPython(`_Node(_code).has_variable('has_at')`)) })
```

Your `has_at` variable should have the value `'@' in email`.

```js
({ test: () => assert(runPython(`
node = _Node(_code).find_variable('has_at')
not node.is_empty() and (
  node.is_equivalent("has_at = '@' in email")

> )

Reviewer 2: `not node.is_empty() and` is dead code in every hint that uses it. `find_variable` returns `Node()` when the variable is missing, and `is_equivalent` short-circuits on that itself:

```python
def is_equivalent(self, target_str):
    # Setting the tree to None is used to represent missing elements.
    if self.tree == None:
        return False
```

Confirmed against the bundled 9.0.1 helpers: for a missing variable the guarded and unguarded forms return the same value. Nine hints in this workshop already use the one-line form (steps 1, 3, 10, 11, 12, 13, 14, 15), so the guard is both redundant and an internal inconsistency. Steps 5 and 6 additionally wrap the single call in an extra set of parentheses.

The same removal applies to: step 4 L36, step 6 L28, step 7 L33, step 8 L33, step 9 L37, step 10 L64, step 14 L40, step 15 L35, step 16 L35, step 17 L35 and L50, step 18 L35, step 20 L32 and L47.

```suggestion
_Node(_code).find_variable('has_at').is_equivalent("has_at = '@' in email")
```
